src/jit/optcse.cpp

```diff
@@ -880,39 +972,103 @@ void Compiler::optValnumCSE_InitDataFlow()
  */
 class CSE_DataFlow
 {
-    BitVecTraits* m_pBitVecTraits;
-    EXPSET_TP     m_preMergeOut;
+    Compiler* m_comp;
+    EXPSET_TP m_preMergeOut;
+    EXPSET_TP m_callKills;
 
 public:
-    CSE_DataFlow(Compiler* pCompiler) : m_pBitVecTraits(pCompiler->cseTraits), m_preMergeOut(BitVecOps::UninitVal())
+    CSE_DataFlow(Compiler* pCompiler)
+        : m_comp(pCompiler), m_preMergeOut(BitVecOps::UninitVal()), m_callKills(BitVecOps::UninitVal())
     {
+        m_callKills = BitVecOps::MakeEmpty(m_comp->cseLivenessTraits);
+        for (unsigned inx = 0; inx < m_comp->optCSECandidateCount; inx++)
+        {
+            unsigned CseAvailBit = inx * 2;
+
+            // a one preserves availability and a zero kills the availability
+            // we generate this kind of bit pattern:  101010101010
```

Comment:
If I understand correctly, these two bits really reflect 3 states:
- The CSE is available and is not killed by a call - `11`
- The CSE is available, but is killed by a call - `10`
- The CSE is not available - `00`

The `10` pattern seems counter-intuitive as an initial setting. If you want to initialize it to "full" it would seem logical to turn on both bits, and if you want to initialize it to "empty" it would seem logical to turn off both bits. What am I missing here?

Reply:
Yes, your understanding is basically correct, but I would express it this way:

`11` - The CSE is available, and is also available when considering calls as killing availability.
`10 `- The CSE is available, but is not available when considering calls as killing availability.
`00`  - The CSE is not available
`01 `- Illegal combination

This code constructs the mask `m_callKills` which is used below to clear the lower bit, - `10` of the `bbCseIn` for every CSE candidate, when the block contains a callsite:

```
            if (block->bbFlags & BBF_HAS_CALL)
            {
                BitVecOps::IntersectionD(m_comp->cseLivenessTraits, cseIn_withCallsKill, m_callKills);
            }
```

Reply:
I can rename `m_callKills` to `m_callKillsMask` to make this more clear.